src/modules/Packages.rb

```diff
@@ -836,10 +841,25 @@ def Reset(keep)
         end
       end
 
+      # special handling for packages, there might be tens of thousands packages
+      # so we need to do it more effectively
+      if keep.include?(:package)
+        # first query the selected packages, then get the details for each selected
+        # package separately to avoid processing a huge list at once
+        # (true = names only without version)
+        Pkg.GetPackages(:selected, true).each do |pkg|
+          Pkg.ResolvableProperties(pkg, :package, "").each do |resolvable|
+            next if resolvable["status"] != :selected || resolvable["transact_by"] == :solver
+            restore_packages << pkg
```

Comment:
I think the fix can be made simpler if we drop `restore_packages` and say `restore << [pkg, :package]` here

Reply:
I deliberately avoided that, there might be thousands of packages, creating extra nested array (a pair) `[pkg, :package]` for each package will take additional memory. Having a simple `String` value instead of nested `Array` might make some memory difference.

(Though I do not have any real numbers. AFAIK the symbols are shared in Ruby so that should not be a big problem, but I'm not sure how Ruby `Array` is effective, esp. if it contains just two items.)